src/Adapter/Product/Update/ProductDuplicator.php

```diff
@@ -527,64 +735,100 @@ private function duplicateTags(int $oldProductId, int $newProductId): void
     }
 
     /**
      * @param int $oldProductId
      * @param int $newProductId
+     * @param array $combinationMatching
+     * @param ShopConstraint $shopConstraint
      *
      * @throws CannotDuplicateProductException
      * @throws CoreException
      */
-    private function duplicateDownloads(int $oldProductId, int $newProductId): void
+    private function duplicateImages(int $oldProductId, int $newProductId, array $combinationMatching, ShopConstraint $shopConstraint): void
     {
-        /* @see Product::duplicateDownload() */
-        $this->duplicateRelation(
-            [Product::class, 'duplicateDownload'],
-            [$oldProductId, $newProductId],
-            CannotDuplicateProductException::FAILED_DUPLICATE_DOWNLOADS
-        );
-    }
+        $oldImages = $this->getRows('image', ['id_product' => $oldProductId], CannotDuplicateProductException::FAILED_DUPLICATE_IMAGES);
+
+        $imagesMapping = [];
+        $fs = new Filesystem();
+        foreach ($oldImages as $oldImage) {
+            $oldImageId = new ImageId((int) $oldImage['id_image']);
+            $newImage = $this->productImageRepository->duplicate($oldImageId, new ProductId($newProductId), $shopConstraint);
+            if (null === $newImage) {
+                continue;
+            }
 
-    /**
-     * @param int $oldProductId
-     * @param int $newProductId
-     * @param array $combinationImages
-     *
-     * @throws CannotDuplicateProductException
-     * @throws CoreException
-     */
-    private function duplicateImages(int $oldProductId, int $newProductId, array $combinationImages): void
-    {
-        /* @see Image::duplicateProductImages() */
-        $this->duplicateRelation(
-            [Image::class, 'duplicateProductImages'],
-            [$oldProductId, $newProductId, $combinationImages],
-            CannotDuplicateProductException::FAILED_DUPLICATE_IMAGES
-        );
+            $newImageId = new ImageId((int) $newImage->id);
+            $imageTypes = $this->productImageRepository->getProductImageTypes();
+
+            // Copy the generated images instead of generating them is more performant
+            foreach ($imageTypes as $imageType) {
+                $fs->copy(
+                    $this->productImageSystemPathFactory->getPathByType($oldImageId, $imageType->name),
+                    $this->productImageSystemPathFactory->getPathByType($newImageId, $imageType->name)
+                );
+            }
+
+            // Also copy original
+            $oldOriginalPath = $this->productImageSystemPathFactory->getPath($oldImageId);
+            $newOriginalPath = $this->productImageSystemPathFactory->getPath($newImageId);
+            $fs->copy(
+                $oldOriginalPath,
+                $newOriginalPath
+            );
+
+            // And fileType
+            $fs->copy(
+                dirname($oldOriginalPath) . '/fileType',
```

Comment:
what is this "/fileType" dir, is it ok to have it hardcoded here? Maybe you should inject it or so?

Reply:
Found it, it comes from the ImageManager::resize method https://github.com/PrestaShop/PrestaShop/blob/df34947093951850bfb53c06d6a5906a8bf90a0e/classes/ImageManager.php#L334